Feat : Rates in BuyNow Modal

bytesaparna opened the pull request:
# Motivation

Rates  added in the marketplace did not show up during purchase of NFT.

# Implementation

Created a hook "useGetRate" to get the data of the rate (like rateType, value,description) and used it in the BuyNowModal. Imported sumCoins function to add the coins. 
Created function named shortenString to shorten the length of error.
Show rates in BuyNowModal.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

- **New Features**
	- Introduced a new pricing structure in the Buy Now modal, dynamically calculating rates based on user input.
	- Added a custom hook, `useGetRate`, for fetching rate information.
	- Implemented a toast notification system for error handling during data mutations.
	- Enhanced the component hierarchy to include React Query for improved data management.

- **Bug Fixes**
	- Enhanced error messages in toast notifications for better user experience.

- **Documentation**
	- Updated internal documentation to reflect new hooks and components.

- **Chores**
	- Added new dependencies to support enhanced functionality.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

bytesaparna committed:
feat:Rates in BuyNow Modal

leifasorensen commented on app/providers.tsx:
```diff
@@ -1,10 +1,13 @@
 import { ApolloProvider } from "@apollo/client";
 import { CacheProvider } from "@chakra-ui/next-js";
 import { ChakraProvider } from "@chakra-ui/react";
+import { QueryClient } from "@cosmjs/stargate";
```
Dependency updates and removals are outside of the scope of this task, and will be omitted from current change requests. These suggestions have been logged to be performed later.

leifasorensen commented on package.json:
```diff
@@ -17,11 +17,13 @@
     "@chakra-ui/icons": "^2.1.1",
     "@chakra-ui/next-js": "^2.1.5",
     "@chakra-ui/react": "^2.8.1",
+    "@cosmjs/amino": "^0.32.4",
```
Dependency updates and removals are outside of the scope of this task, and will be omitted from current change requests. These suggestions have been logged to be performed later.

leifasorensen commented on src/lib/react-query/client.ts:
```diff
@@ -0,0 +1,36 @@
+import { QueryClient } from "@tanstack/react-query";
+
+
+const reactQueryClient = new QueryClient({
+    defaultOptions: {
+        queries: {
+            staleTime: 5 * 60 * 1000
+        },
```
This seems an acceptable conflict as applied rates should not be changing during active sales.

leifasorensen commented on src/lib/andrjs/hooks/ado/useGetRate.ts:
```diff
@@ -0,0 +1,37 @@
+import useAndromedaClient from "@/lib/andrjs/hooks/useAndromedaClient";
+import { useQuery } from "@tanstack/react-query";
+
+export interface Rate {
+    "local": {
+        "rate_type": "additive" | "deductive",
+        "value": {
+            "flat"?: {
+                "denom": string,
+                "amount": string
+            },
+            "percent"?: {
+                "percent": string
+            }
+        },
+        "decription": string
+    }
+}
```
The rate_type value cannot be empty as it is enforced from the contract as required. This means undefined values would only be present for data retrieval or reference failures, which make the dual undefined value concern irrelevant. This does not need to be addressed.

Resolving the typo should be sufficient for resolving this request.

bytesaparna committed:
fix typo error and conflict related to rate type

leifasorensen committed:
Updated client.ts error handling

- Added proper type annotation for the error parameter (`err: unknown`)
- Improved the error handling logic to properly type-check and handle different error formats
- Created a proper Error object regardless of the input error type
- Fixed the type safety issues while maintaining the same functionality
